h/services/user_signup.py

```diff
@@ -72,38 +83,62 @@ def signup(self, require_activation: bool = True, **kwargs) -> User:  # noqa: FB
                 user=user,
                 provider=i_args["provider"],
                 provider_unique_id=str(i_args["provider_unique_id"]),
+                email=i_args.get("email"),
+                name=i_args.get("name"),
+                given_name=i_args.get("given_name"),
+                family_name=i_args.get("family_name"),
             )
             for i_args in identities
         ]
 
         self.session.add(user)
+        try:
+            self.session.flush()
+        except IntegrityError as err:
+            # Concurrent, conflicting signup requests can all pass validation
+            # (each request finding that no matching user exists in the DB)
+            # and go on to try to insert conflicting users. When this happens
+            # one of the requests successfully inserts a user and the others
+            # fail here with an IntegrityError.
+            #
+            # Raise different exceptions depending on which column was in
+            # conflict because this is of interest to exception handlers in the
+            # calling code.
+            #
+            # But note that if this code raises an exception about one column
+            # being in conflict that doesn't mean that other columns were not
+            # *also* in conflict: two requests can try to insert users with the
+            # same email address, username *and* identity.
+            #
+            # As soon as the first integrity check fails Postgres raises an
+            # error. It doesn't continue to process other constraints to see
+            # what other checks would have also failed.
+            if (
+                'duplicate key value violates unique constraint "uq__user__email"'
+                in err.args[0]
+            ):
+                raise EmailConflictError from err
+
+            if (
+                'duplicate key value violates unique constraint "ix__user__userid"'
+                in err.args[0]
+            ):
+                raise UsernameConflictError from err
+
+            if (
+                'duplicate key value violates unique constraint "uq__user_identity__provider"'
+                in err.args[0]
+            ):
+                raise IdentityConflictError from err
+
+            raise  # pragma: no cover
```

Comment:
The final `raise` here should never happen: no other type of `IntegrityError` is expected. The `raise` is here so that if an unexpected `IntegrityError` _does_ somehow get raised we don't silence it. But I decided not to add test coverage: since there are no other constraints to be violated here it's not possible to construct a test that would hit this line except by mocking SQLAlchemy.